db/migrate/20251216104733_create_member_roles.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+class CreateMemberRoles < ActiveRecord::Migration[8.0]
+  disable_ddl_transaction!
+
+  def up
+    unless index_exists?(:memberships, [:id, :organization_id], name: "membership_fk_idx")
+      add_index :memberships, [:id, :organization_id], unique: true, name: "membership_fk_idx", algorithm: :concurrently
+    end
+
+    unless table_exists?(:member_roles)
+      create_table :member_roles, id: :uuid do |t|
+        t.uuid :organization_id, null: false
+        t.uuid :membership_id, null: false
+        t.references :role, type: :uuid, null: false
+        t.boolean :admin, default: false, null: false
+        t.datetime :revoked_at
+        t.timestamps
+
+        t.index [:membership_id, :role_id], unique: true, where: "revoked_at IS NULL", name: "unique_member_role_idx"
+        t.index :membership_id, name: "unique_member_role_membership_idx"
+        t.index [:membership_id, :organization_id], name: "member_role_membership_fk_idx"
+        t.index [:role_id, :admin], name: "member_role_role_fk_idx"
+
+        t.foreign_key :roles,
+          column: [:role_id, :admin],
+          primary_key: [:id, :admin],
+          name: "member_role_role_fk"
+        t.foreign_key :memberships,
+          column: [:membership_id, :organization_id],
+          primary_key: [:id, :organization_id],
+          name: "member_role_membership_fk"
+      end
+    end
+
+    safety_assured do
+      execute <<~SQL.squish
+        CREATE OR REPLACE FUNCTION ensure_member_role_consistency() RETURNS TRIGGER AS $$
+          BEGIN
+            IF TG_OP = 'INSERT' THEN
+              IF EXISTS (
+                SELECT FROM roles
+                WHERE id = NEW.role_id AND COALESCE(organization_id, NEW.organization_id) != NEW.organization_id
+              ) THEN
+                RAISE EXCEPTION 'Cannot assign a role to a membership in a different organization';
+              END IF;
+            ELSIF TG_OP = 'DELETE' THEN
+              RAISE EXCEPTION 'Hard deletion of role membership is not allowed; use soft deletion instead';
+            ELSIF OLD.revoked_at IS NOT NULL THEN
+              RAISE EXCEPTION 'Revoked role membership cannot be modified';
+            ELSIF OLD.organization_id != NEW.organization_id THEN
+              RAISE EXCEPTION 'Role membership cannot be moved to a different organization';
+            ELSIF OLD.role_id != NEW.role_id THEN
+              RAISE EXCEPTION 'Role membership cannot be moved to a different role';
+            ELSIF OLD.membership_id != NEW.membership_id THEN
+              RAISE EXCEPTION 'Role membership cannot be moved to a different membership';
+            END IF;
+            
+            RETURN NEW;
+          END;
+        $$ LANGUAGE plpgsql;
+
+        CREATE TRIGGER ensure_member_role_consistency
+        BEFORE INSERT OR UPDATE OR DELETE ON member_roles
+        FOR EACH ROW EXECUTE FUNCTION ensure_member_role_consistency();
+      SQL
+    end
```

Comment:
I'm in favor of removing the entire function (and other `ensure_*_consistency` functions) and keeping the app responsible for this.

Reply:
I understand what you mean but thihnk today we don't want to add all these constraint in the BD

Reply:
fixed